runtime/runtime/src/congestion_control.rs

```diff
@@ -462,3 +488,32 @@ fn overflow_storage_err() -> StorageError {
 fn safe_add_gas_to_u128(a: u128, b: Gas) -> Result<u128, IntegerOverflowError> {
     a.checked_add(b as u128).ok_or(IntegerOverflowError {})
 }
+
+/// Each shard receieves receipts from all other shards, but it could happen
+/// that all of the sender shards sends a lot of data, which would cause
+/// the state witness to be really large. To prevent this, the usual limit
+/// on the size of outgoing receipts is pretty small (outgoing_receipts_usual_size_limit).
+/// However, on each block height, one of the sender shards is allowed to send more data
+/// (outgoing_receipts_big_size_limit) to the receiver shard.
+/// This allows large receipts to be sent when its their turn, without overloading
+/// the receiver shard.
+///
+/// This function is responsible for pairing the sender shard that will be allowed to send
+/// more with the receiver shard that will receive the extra data.
+/// Each sender shard is allowed to send more data to one "special" receiver shard.
+/// Each receiver shard will receive extra data from only one sender shard.
+pub(crate) fn choose_special_shard(
+    sender_shard: ShardId,
+    block_height: BlockHeight,
+    num_shards: usize,
+) -> Option<ShardId> {
+    if num_shards == 0 {
+        return None;
+    }
```

Comment:
This seems off. Can you clarify under what circumstances this may happen? Also if we can avoid it this would be great but I can't tell without context.

Reply:
Yeah I'll remove it, it's a remnant of the previous version where special shard couldn't be the same as the source shard :+1:

Reply:
On second thought, `num_shards` comes from `ApplyState::congestion_info`, and I'm not 100% sure that it'll always have nonzero length. It's safer to assume that it can be zero, and just not choose a special shard then.

```rust
let num_shards = apply_state.congestion_info.len();
let special_shard: Option<ShardId> =
    choose_special_shard(apply_state.shard_id, apply_state.block_height, num_shards);
```